### ISSUE TITLE
Implement support for disabling rules as local as possible.

### ISSUE THREAD

**Issue by maxim-tschumak** [zalando-incubator/zally#446]
**As** a user of Zally
**I want** to be able to disable specific rules on different levels of the Swagger hierarchy.
**So that** the integration of Zally into the build process is more flexible

Related to / Triggered by: #406

Example by @tfrauenstein:
> For instance, rule "Should: Limit number of Sub-Resource Levels" only applies to resource paths and disabling should be for a specific path -- technically an OpenAPI "path item object" instance -- and not for the complete API. Hence disabling should only become effective, if it is done within the smallest context of the OpenAPI element relevant for the rule. Is this requirement agreed?

**Comment by innokenty** [comment]
I don't see any other way of implementing this, rather than hardcoding checks of the `x-zally-ignore` extension checks in the code of every single rule on each level of swagger doc hierarchy. That's a quite time-consuming approach, also taking into consideration all the possible bugs and combinations of different inner tags in there.

This is doable though, but I would first wait and make sure this is actually a highly demanded feature. I doubt there are many people that really want to disable some checks on some level and not globally.

**Comment by maxim-tschumak** [comment]
I see two options, besides the one proposed by @innokenty :

1. A dynamic **RuleRunConfiguration** would describe against which _parts_ of the API (meta information, concrete endpoints, concrete models, concrete parameters) the rule should check. Each rule has to implement an interface to check specific parts of the API (e.g. `Rule#checkEndpoint(e string): List<Violations>`).

2. **Exclude** paths marked with `x-zally-ignore` **after the check**.

@netme @maschleg @tkrop @mfellner What do you think?

**Comment by tfrauenstein** [comment]
yes, implementation is more involved. 
However, I see it as a must feature already in scope of the original feature idea:

"When drafting an API, the API designer usually runs Zally multiple times.
He then checks Zallys violation output and decides how to handle each violation.
Sometimes API is changed as to avoid the violation, sometimes API is not changed on purpose -- often the case for SHOULD and MAY violations.
Improve API designer convenience via support of pragma directives inside API definitions that disable checking of specific rules for specific OpenAPI elements.
By this, API designers can avoid efforts for redundant checking of violations for multiple Zally runs during API evolution. It is also a must-feature for integration of Zally in automated CD or operation monitoring tooling."

I'm sceptical that 1st delivery step of only global disabling really make sense: 
Disabling should be local and an exception, global disabling mutates a specific exception to a general ignore rule default and has the danger of contrary effects to the Zally purpose of improving guideline compliance. During API evolution more and more rules are disabled and that these rule check exceptions are also applied for API extensions during API evolution life cycle without being checked by API designer and confirmed on purpose. 

In any case, applying of disabling should be explicitly reported as Zally output hint. Do we already support this?

**Comment by maxim-tschumak** [comment]
Agree! I've created an issue to inform the user about the disabled rules (#485).

**Comment by roxspring** [comment]
I wonder if "as local as possible" is being taken too literally as if to mean at any point in the swagger definition that vendor extensions are allowed. Maybe we could make more progress if we were to define a few specific high priority locations that should be supported.

For example if path level ignoring is really important then we could add a new SwaggerPathRule base class that embeds the path level ignoring logic and allows subclasses to implement a `validate(swagger: Swagger, pattern: String,path: Path)` method. Similar could be added for other important levels, but with each addition adding more potential confusion over which base class to extend.
